## Introduction {.page_break_before}

The gene which consists of essentially DNA is an inheritable unit to daughter cells. DNA wraps around the proteins called histones, and six histones specifically aggregate and the DNA wrap 2 times around the histone aggregates called nucleosomes and nucleosomes plus DNA tangles like a phone cord. In each loop there are six nucleosomes, at the end the resulted tangled loops are named chromatins. By the time depending on the job of genes or gene elements (transcription elements such as regulators, enhancers, silencers etc.), histone tails are post-transitionally modified and the locations of those histone marks widely varies throughout the genome wide. Not only the localizations of these histone markers but also their varieties and the effect of each different post translational modification on the function of transcription and replication. Genes are localized on special positions on genome called loci and their expression result into production of RNA and sometimes proteins as well with processes called transcription and translation, sequentially. These end-products that are coming from either transcription or translation are key biological entities for cell growth, development, metabolism and its survival. 

With respect to the activity of genes, while some portion of DNA are euchromatin (open chromatin) other portions of DNA can be heterochromatin (close chromatin). Euchromatin regions are found at sites where housekeeping genes or frequently demanded genes localized and their structure are available for proteins to interact with the DNA. On the other hand, heterochromatin regions found at places where the genes are not frequently expressed and hence it is very difficult for proteins to reach at these positions. The changes of DNA as oppose to its origin is called mutations or variations depending on its impact and frequency of prevalence in the populations. There are some changes also occur within the organism but not in DNA sequence which are called epigenetic alterations and these changes are known to affect critically the function and activity of genes as well. The key challenge for a cell to function appropriately depends on its ability to continue its life cycle without mistakes. This is because completing the life cycle of a cell with a mutation found on the DNA sequence without detection by cell cycle checkpoints lead that mutated DNA to be inherited by daughter cells thereby maintenance and multiplication of abnormal DNA material ends with formation of called tumor cells. The integrity of genome sustained by DNA repair mechanisms. All DNA repair mechanisms coincide at a common event which is identification / recognition of the DNA damage. Together recognition and correction of damage part of the DNA is called DNA damage repair (DDR). DNA damage recognition events are usually having common steps and same order. DDR begins with detection of mutation site on the genome by one or two proteins and follows by recruitment and accumulation of complex of proteins to the site of damage for replacing wrong nucleotides with the correct ones. The damage recognition phase is the key determiner in other words rate limiting step of the DNA repair mechanisms. Any impairments of DDR which are satisfied by complex proteins interacting with DNA and other protein(s) depending on the type of damage repair mechanism if not catch at cell cycle checkpoints end up with defectiveness of cell function in best scenario or severe diseases.

There are two classes of DNA repair pathways and they further divides into sub-mechanisms. Two classes of DNA repair pathways are damage repair due to polymerase errors after translation or replication and after external related DNA damage formation repairs. The former includes two repair mechanisms called a proofreading and a mismatch repair. The proofreading is a process which DNA polymerase checks each nucleotide added during elongation processes to avoid errors. Moreover, mismatch repair is the process starts after proofreading is finished and evolved based on fixing the remaining unfixed wrong nucleotides. The reason of proofreading does not work at complete efficiency is because polymerase could slip during elongation and may add wrong nucleotides to newly synthesized strand of DNA. On the other hand, externally induced DNA damage formation repairs occur due to chemical or physical stress related factors. A chemical reversal, base excision, nucleotide excision and double stranded break repair are sub pathways of latter classes of DNA repair that is described above. 

Briefly, the chemical reversal is elimination of extra chemical groups to DNA backbone or DNA bases. Double stranded break repair is divided into two mechanisms as non-homologous end joining (NHEJ) and homologous recombination (HR). After double stranded DNA cleaved at both strands and yield double stranded break (DSB) if there is homolog donor strand of DNA, involving original sequence of cleaved portion, and presents inside the cell, the HR occurs and DSB is fixed using homolog strand. If homolog donor DNA does not exist inside the cell in order to instantly fix the DSB non homologous end joining undergoes by directly ligating two ends of double stranded DNA together and resulting into the insertion of indel mutations. Like mismatch repair, the base excision repair is and after detection of wrongly incorporated nucleotide(s) is replaced with the correct one. Lastly, nucleotide excision repair (NER) is categorized into two mechanisms named NER global genome repair (GR) and NER transcription coupled repair (TCR). These specifically evolved repair systems are fixing the UV lesions either 6-4 pyrimidone pyrimidine photo products ((6-4)PPs), cyclobutene pyrimidine dimers (CPD) or both. The two of the nucleotide excision repairs are evolved for correcting dimerization because of the exposure of period of time to the UV light. 

NER differs than other damage repair mechanisms in terms of the replacement of wrongly incorporated bases with the new ones. After the 
detection of dimers either by global repair or transcription coupled repair, the dimerized bases are discarded together with varying degree of its surrounding bases depending on at which organism the process happens. TCR and GR are known to be shown different repair rates. As TCR repair much more efficiently and rapidly than GR, both are found to be more effective in repairing of 6-4PPs than CPDs. 
Moreover, while TCR works only at transcribed strand, GR is revealed that could function at both strands. The UV damage recognition event in TCR is performed with RNA polymerase itself therefore, TCR better than GR in terms of effectiveness and rapidness since damage identification is the rate limiting step. Furthermore, there are two recognition damage recognition factors, DDB2 and XPC, takes a role in GR. Verificationevent of GR is handled with the help of repair factor XPA via gathering pre-excision complex which consists of TFIIH complex to damage site. TFIIH transcription factor is a repair complex of proteins formed together with XPD, XPB, ERGG1, XPF, XPA, RPA, XPG, p62, p34, p44, p52 and TTDA. The recognized and verified site then proceed to the repair that endonuclease XPF and XPG cleave the damage region and its surroundings and the newly formed gap is ligated with DNA Polymerases. 

The generation and improvement of sequencing technologies recently ease the creation new methodologies of sequencing DNA regions for 
different purposes. A xr-seq and a damage-seq are one of the sequencing methodologies for sequencing the sequences where UV damage repair happens and sequencing UV damages, sequentially. The xr-sequencing is a method captures the portion of reads by 5â€™ adaptors that is cleaved off from genome for replacing with the correct nucleotides before its degradation. Then, damage reversal of captured amplicons is followed by the PCR amplification of amplicons and amplified amplicons are purified and sequenced. The outcome of xr-seq does not provide particular damage location rather gives un-dimerized DNA portion. Damage-seq was developed for detecting the UV damage throughout the genome and has similar steps with xr-sequencing technique except after ligation of first adaptor and immunoprecipitation instead of damage reversal, purification of primer extended sequence and second adaptor ligation is performed. Moreover, the damage-seq data in fact, do not include the position of UV damage (dimerization), the specific location of damage is obtained during data preparation. The UV and cisplatin induced samples have been cultured and xr-seq and damage-seq technologies were applied, sequentially. The xr-seq showed unequally distributed NER pattern throughout the genome and to answer the question of whether this pattern is because of the NER mechanisms or heterogeneity of UV damage exposure. 

The damage-seq techniques were also performed on the same UV exposed biological sample and the findings eliminated the plausibility of 
heterogeneity of UV damage distribution thorough the genome. This is because damage-seq experiment have been demonstrated that the UV 
damage distribution was following homogeneous pattern all along the genome as oppose to NER. Therefore, it was stated that the 
preferential UV damage repair should be because of NER itself. Even though reasons of the pattern of UV damage repair could be guessed, 
the exact answers still need researches for elucidation. 

Here, we aimed to provide an answer not completely but partially for fulfilling this uncertainty and for being pioneer to upcoming 
other researches related with the gap. We have selected and applied machine learning algorithm FFFF using 15 histone markers with trial 
and error to be able to analyze the influence of histone markers and chromatin states on NER. We found that a normalization method of 
count numbers after the reference genome mapping RRRR works better than YYYY calling in terms of predicting high and low DNA repair 
profiles with given histone markers. Secondly, for same cell line type and mutation types, we succeed to predict DNA repair profiles within specified region (5000 base pair) of genome without looking at actual DNA repair pattern. Histone markers aaa, bbb, ccc found significantly related with DNA UV damage global repair mechanisms. These findings provide knowledge for explaining the cause of discordant pattern of UV damage and repair, even though results alone do not enough to clarify the reason of difference of repair and damage. Together with the results, it is now revealed that the histone markers are one of the main determiners of DNA repair heterogeneity.
